### PR TITLE
Process data in model

### PR DESCRIPTION
**Pull request by abiola-arobieke**
In this milestone:

The user, post, comment, and likes models were created.
An association between models was set up.
Models were used to insert data using the rails console.
Custom methods were created for each model.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: :author
+  has_many :comments, foreign_key: :author
+  has_many :likes, foreign_key: :author
```
Comment (rloterh) [inline app/models/user.rb]: - [x] Let's ensure that foreign keys are defined appropriately, like below: 
```ruby
# app/models/user.rb
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+
+  def update_posts_counter
+    author.update(posts_counter: author.posts.count)
+  end
```
Comment (rloterh) [inline app/models/post.rb]: - [x] [OPTIONAL]
It's a good practice to use one of the Rails Active Callbacks `after_save` to call the `update_post_counter` method every time a new post has been saved to the database, and when `after_save` is called mark the `update_post_counter` method as _private_ accordingly, which should be similar to:

```ruby
 # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post
+
+  def update_comments_counter
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (rloterh) [inline app/models/comment.rb]: - [x] [OPTIONAL]
Similarly, it's a good practice to use one of the Rails Active Callbacks `after_save` to call the `update_comment_counter` method every time a new comment has been saved to the database, and when `after_save` is called mark the `update_comment_counter` method as _private_ accordingly, which should be similar to:

```ruby
 # app/models/comment.rb
  after_save :update_comment_counter

  private

  def update_comment_counter
    author.increment!(:comments_counter)
  end
```

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  has_many :author, class_name: 'User', foreign_key: 'author_id'
+  has_many :post
+
+  def update_likes_count
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (rloterh) [inline app/models/like.rb]: - [x] [OPTIONAL]
Again, it's a good practice to use one of the Rails Active Callbacks `after_save` to call the `update_like_counter` method every time a new like has been saved to the database, and when `after_save` is called mark the `update_like_counter` method as _private_ accordingly, which should be similar to:

```ruby
 # app/models/like.rb
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Abiola, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🏁 🚦 ⛪ 🚜 🏃 🌴 🏫 🌳 🏘️

## Highlights 👍  
- Good job so far with creating model associations!
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_Kindly consider my inline comments marked **[OPTIONAL]** under this review_

**You can also consider:**
Let's improve the look of the ReadMe by removing the underlined:

image

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/like.rb (hunk 4 above)
Comment (abiola-arobieke) [reply app/models/like.rb]: Thank you for the suggestions.

**Review by danmainah — APPROVED** [review]
Hi @abiola-arobieke , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 

## Highlights
 ✔️  Implemented the change requested to ensure the has_many relationship in the user model is correctly defined.
✔️  Implemented the change requested to utilize the benefits in  Rails Active Callbacks after_save in the custom model methods.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**